runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -107,10 +140,26 @@ synchronized ActivateWorkResult activateWorkForKey(ShardedKey shardedKey, Work w
       return ActivateWorkResult.EXECUTE;
     }
 
-    // Ensure we don't already have this work token queued.
-    for (Work queuedWork : workQueue) {
-      if (queuedWork.getWorkItem().getWorkToken() == work.getWorkItem().getWorkToken()) {
+    // Check to see if we have this work token queued.
+    Iterator<Work> workIterator = workQueue.iterator();
+    while (workIterator.hasNext()) {
+      Work queuedWork = workIterator.next();
+      if (queuedWork.id().equals(work.id())) {
         return ActivateWorkResult.DUPLICATE;
+      } else if (queuedWork.id().cacheToken() == work.id().cacheToken()) {
+        if (work.id().workToken() > queuedWork.id().workToken()) {
+          removeIfNotActive(queuedWork, workIterator, workQueue);
+          workQueue.addLast(work);
+          return ActivateWorkResult.QUEUED;
+        } else {
+          return ActivateWorkResult.STALE;
+        }
+      } else if (queuedWork.id().workToken() == work.id().workToken()) {
+        if (queuedWork.id().cacheToken() != work.id().cacheToken()) {
```

Comment:
this is known true from previous statement in else if chain

I think the case 
queuedWork.id().workToken() == work.id().workToken()
could just be removed though and fall through to the bottom. If the cache tokens are different we're not actually sure which is the valid item in the backend.  And in that case it is safer to keep both since then we will eventually complete the right one.

Reply:
done